include/osquery/expected.h

```diff
@@ -0,0 +1,138 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <memory>
+#include <osquery/error.h>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+/** Utility class that should be used in function that return
+ * either error or value. Expected enforce developer to test for success and
+ * check error if any.
+ *
+ * enum class TestError { SomeError = 1, AnotherError = 2 };
+ * Expected<std::string> function() {
+ *   if (test) {
+ *    return "ok";
+ *   } else {
+ *    return std::make_shared<Error<TestError>>(TestError::SomeError);
+ *   }
+ * }
+ *
+ * Expected:
+ * ExpectedUnique<PlatformProcess> function() {
+ *   if (test) {
+ *    return std::make_unique<PlatformProcess>(pid);
+ *   } else {
+ *    return std::make_shared<Error<TestError>>(TestError::AnotherError);
+ *   }
+ * }
+ *
+ * auto result = function();
+ * if (result) {
+ *   ...use *result
+ * } else {
+ *   auto error = result->getError();
+ * }
+ */
+
+namespace osquery {
+
+template <class T>
+class Expected {
+ public:
+  Expected(T object) : object_(std::move(object)), hasError_(false) {}
+  Expected(ErrorBase* error) = delete;
+  Expected() : error_(nullptr), hasError_(false) {}
+  Expected(std::shared_ptr<ErrorBase> error)
+      : error_(std::move(error)), hasError_(true) {}
+  Expected(std::unique_ptr<ErrorBase> error)
+      : error_(std::move(error)), hasError_(true) {}
+  template <class ErrorT>
+  Expected(std::shared_ptr<Error<ErrorT>> error)
+      : error_(std::static_pointer_cast<ErrorBase>(error)), hasError_(true){};
+
+  Expected(const Expected&) = delete;
+
+  ~Expected() {
```

Comment:
Either make the destructor `virtual` or whole class `final`. It is nothing, but without it `descendants` would't call destructor of base class.